filament/backend/src/metal/MetalHandles.h

```diff
@@ -112,12 +112,24 @@ class MetalSwapChain : public HwSwapChain {
     void* frameCompletedUserData = nullptr;
 };
 
+class MetalBufferObject : public HwBufferObject {
+public:
+    MetalBufferObject(MetalContext& context, uint32_t byteCount);
+
+    void updateBuffer(void* data, size_t size, uint32_t byteOffset);
+    const std::shared_ptr<MetalBuffer>& getBuffer() const { return buffer; }
+
+private:
+    uint32_t byteCount;
+    std::shared_ptr<MetalBuffer> buffer = nullptr;
+
+};
+
 struct MetalVertexBuffer : public HwVertexBuffer {
     MetalVertexBuffer(MetalContext& context, uint8_t bufferCount, uint8_t attributeCount,
             uint32_t vertexCount, AttributeArray const& attributes, bool bufferObjectsEnabled);
-    ~MetalVertexBuffer();
 
-    std::vector<MetalBuffer*> buffers;
+    std::vector<std::shared_ptr<MetalBuffer>> buffers;
```

Comment:
Do we really need shared_ptr here instead of unique_ptr? Is MetalBufferObject the only other place where those pointers are used? Can a MetalbufferObject outlive this vector? If not, it doesn't look like we need shared_ptr.

Reply:
In the future, we might be able to simplify the Driver API and make it so that buffers are always owned by a HwBufferObject.  This would let us remove the shared ownership.

Vulkan is also using shared_ptr (for now) because each buffer in this list was created either by a HwVertexBuffer or by a HwBufferObject, and it lets you avoid the ownership bookkeeping.

Reply:
Yeah, what Philip said. If we refactor Filament's VertexBuffer to always use BufferObjects under the hood (and I think we should), then we won't need the additional bookkeeping here.